### PR TITLE
786 SegString as single entry (commits grouped by type) & Complex Object in Symbol Table

### PR DESCRIPTION
**Pull request by glitch**
## Update - Typing Hierarchy added for Symbol Table & Complex Objects
UPDATE: I added the changes for the Symbol Table / SymEntry Type Hierarchy which allow us to put different types of entries into the SymbolTable.

I consolidated Type Hierarchy changes into a single commit [8a6d5082fce6379414874cd757f508288d5db106 ](https://github.com/Bears-R-Us/arkouda/pull/830/commits/8a6d5082fce6379414874cd757f508288d5db106) so you can see them in isolation.

The base of the typing system is the enum:
```chapel
    enum SymbolEntryType {
        AbstractSymEntry,  // Root Type from which all other types will inherit

            TypedArraySymEntry, // Parent type for Arrays with a dtype, legacy->GenSymEntry
                PrimitiveTypedArraySymEntry, // int, uint8, bool, etc.
                ComplexTypedArraySymEntry,   // DateTime, TimeDelta, IP Address, etc.

            CompositeSymEntry, // Parent type for things which are composites of arrays
                SegStringSymEntry,    // SegString composed of offset-int[], bytes->uint(8)
                CategoricalSymEntry,  // Categorical
                GroupBySymEntry,      // GroupBy

            AnythingSymEntry, // Placeholder to stick aritrary things in the map
            UnknownSymEntry,
            None
    }
```

Corresponding to the types are the Entries in previously we had GenSymEntry & SymEntry(type), now we have
* AbstractSymEntry
    * GenSymEntry (to be renamed in future to TypedArraySymEntry)
        * SymEntry(type)  to be renamed in the future to PrimitiveTypedArraySymEntry(type)
    *  CompositeSymEntry
        * SegStringSymEntry 

See `MultiTypeSymEntry.chpl` for  the code.
A number of convenience, factory, and helper procedures were added to make casting between types less obtrusive.

##  Original changes
SegString as a composite SymEntry

Issue #786 
Formerly PR https://github.com/Bears-R-Us/arkouda/pull/819 but with commits grouped by

- server side Chapel changes [342d0a9](https://github.com/Bears-R-Us/arkouda/pull/830/commits/342d0a9b3ca783d60f0e3fda0bd07353fb33ebaa)
- client side Python changes [abd07f5](https://github.com/Bears-R-Us/arkouda/pull/830/commits/abd07f52e8bdd66937801b87743e72c57d5e9065)
- both Chapel & Python unit tests [851067a](https://github.com/Bears-R-Us/arkouda/pull/830/commits/851067af88bbb4b4b4aeb8382e26b2cb83cf6a63)
- benchmark testing [70e3387](https://github.com/Bears-R-Us/arkouda/pull/830/commits/70e33877c59fd09d2c1277b4f4a1258ec07ac57e)

(+1 minor fix and pydoc updates)

#### Main points
- MultiTypeSymEntry.chpl to add `SegStringSymEntry:GenSymEntry` extension for storing another _type_ in the table.
- SegmentedArray.chpl contains 4 factory methods for constructing/assembling/retrieving SegString object from various pieces
- SegmentedMsg.chpl
    - Two new msg cmds for assembling SegString from parts already in SymTab (handles client side Strings creation from arrays) and tondarray to handle passing by the values of the appropriate/component.  Both of these are related to the fact that we can't pass more than one array client <-> server at one time and there is no _state_ regarding messages on the server, so it needs to be command based.
    - Everything else is really updating the former, manual SymEntry creation of the pieces or interaction with them down to the factory method for creating a SegString object and passing back the single entity meta-info for client side construction.
- Most of the Msg related changes echo the previous sub-bullet; aka update how we access a SegString or allocate a new one as well as the msg contents we pass back and forth between the server.
- NumPyDType (see _ugly bits_ below)

#### The ugly bits
- NumPyDType: Added `str`, did we want to do it like this?  This breaks down with something like "Categorical"
- Messaging layer
    - `legacy_placeholder` put in for areas where args are referenced by position (Improve: key=value style arg parsing?)
    - RegistrationMsg (and UniqueMsg) highlights some of the larger issue of fixed number of args both in the GenSymEntry and in the message passing layer, see `st.attrib(id)`
        - Case 1: when we receive msg from client to _do something_ with the object we get passed the two component names joined by `+`, but we don't really need the second name.
        - Case 2: Server response msg for array creation; I use the second position to pass back the  nbytes info because I can't fit it in the fixed number of args of the first one.  We only have 1 entry, but it has one extra meta-info part.
        - `created id str 1 1 (1,) 1+created id str 1 1 (1,) 1` -> `created id str 1 1 (1,) 1+created bytes.size 1234` to pass values.nBytes

#### Main points
- Generally these changes mirror the server-side changes.  Instead of having Strings object with two internal pieces it now has one, so we have to shrink down all of the places which constructed things with two parts or accessed the String _guts_ directly.
    - strings.py
        - Two new static constructors to handle how we make strings objects on the client.  (Come back to this after looking at the init method)
        - Most changes here are the same reduction in the message layer.  Before we passed two names, now we pass one and `legacy_placeholder` (just following the same pattern I did server side so we can do a global search on that string for future udpates).
        - Note on the to_ndarray, mirrors server, we need to get both array components but can only do one at a time right now.
    - Everything else is mostly the message passing send/receive message parsing etc.
- Two main ways to make a strings
    - SymEntry components already exist in SymTab because we created arrays from values on the client and sent them to server, in which case we need to _assemble_ the pieces.
    - We request the server to construct a Strings object on the server-side (i.e. randomStrings funcs, read from hdf, the result of some type of operation like concat) in which case we can used the response message which contains the "joined created+created" message construct.  I put that parsing in its own function client side so we can just pass the response msg.

Note on `Strings.hash` I left this alone, but I'm not sure we need the combined hashes of each component anymore (i.e.  hash1+hash2) maybe we can have the server handle that now, see GroupBy where this is called.

## Common patterns / issues
Regardless of approach and which complex object we try to push server side we run into these common issues.
- We can only pass one component array at a time so there is some assembly required on both the server and client depending on who created the components (i.e. `ak.array(["a", "b", "c"])` vs server `randomStringsMsg`)
- Fixed number of metadata pieces in `GenSymEntry` and `st.attrib(id)` call.  Positional arg parsing makes this tough, either a switch to json or a key=val structure might make this better but I view that as an incremental change and a larger issue.
- This handles composition of pieces reasonably well, but it doesn't address multi-dimensionality.  It's still a collection of 1-D arrays... and maybe that's OK.
- Type casting/coercing from GenSymEntry to the underlying concrete impl.

## Additional changes
- Unit tests in Python
- Unit tests in Chapel
- benchmark tests

### CODE REVIEW

File: src/MultiTypeSymEntry.chpl
Comment (reuster986) [inline src/MultiTypeSymEntry.chpl:30]: Although we had a lot of `try!` in the initial codebase, lately we've been trying to avoid it because it halts the server. Is it possible to get rid of the `try!` and make this a throwing function, or is this a special case where that isn't allowed? I forget the rules...

Comment (glitch) [reply src/MultiTypeSymEntry.chpl:30]: I talked to @mhmerrill about that too.  This follows the current `toSymEntry(type etype)` logic which is a `try!` because I didn't want to deviate from the current behavior (line 52 on master branch).

**Comment by reuster986** [comment]
@glitch I'm going through this now, and I like what you've done!

I've given some thought to the NumPyDType / `str` question, and it might warrant its own issue, but I propose that we decouple the (Gen)SymEntry `dtype` attribute from native Chapel types. I think that the `dtype` should consist solely of the `enum` in NumPyDtype, and should abstract away all Chapel type information, with each child class of `GenSymEntry` encapsulating its own logic for how to determine the Chapel type(s) of the underlying array(s). This would involve changing a lot of code -- for example, `toSymEntry(int)` --> `toSymEntry(DType.Int64)`, but it would allow us to more flexibly add complex data types to the symbol table.

One thing I want to do soon is migrate `ak.Datetime` and `ak.Timedelta` over to the server side. If we decouple dtype from Chapel types, then the dtypes of these classes would be `DType.Datetime64` and `DType.Timedelta64`, respectively, and the classes would handle interpreting the underlying data as Chapel `int` arrays.

Similarly, I'd like to add server-side support for segmented arrays with elements of other than `uint(8)` type. `SegString` would still be its own thing with string-specific methods, but there is already a need for segmented arrays of `int64` in one of the applications I'm targeting, and it does not fit neatly into a Chapel type. But it could easily have a dtype of `DType.SegString_int` or similar.

Anyway, this proposal is outside the scope of this PR, and I'm ok with how it's done here for now.

**Comment by glitch** [comment]
@reuster986 I know exactly what you mean :)  I felt like I was forcing-fitting `str`/`string` into place more out of a concern that something would break if I didn't have one.  In retrospect I think if I just set the same pieces that get set in the `GenSymEntry.init()` call, then I could've probably skipped the native chapel-type aspect altogether.

**Comment by glitch** [comment]
Rebased against master to fix conflict and keep things up to date.

**Comment by glitch** [comment]
I added the changes for the Symbol Table / SymEntry Type Hierarchy which allow us to put different types of entries into the SymbolTable.

I consolidated Type Hierarchy changes into a single commit e691ed4 so you can see them in isolation.

I updated the ticket description to give a brief overview of the Symbol Type hierarchy.

**Comment by glitch** [comment]
@reuster986 I know what you mean.  It really hinges on whether or not we're going to allow arbitrary objects to be put in the SymbolTable i.e. non-array objects (obviously those wouldn't have `shape`, `size`, `dtype` etc.

I have a few thoughts on this:

1. I toyed with the idea of having another layer in the hierarchy that designated things as "Arrays" (it would be a common parent of GenSymEntry & CompositeEntry) but it gets problematic with composites.  SegString has two arrays and they both have different lengths; we kind of squint at it now and use the offsets array as the "real" length, but as the list of Composites grow things like shape, length, dtype get weird.  Would Categorical & GroupBy have a dtype?  (I also got worried about adding too many layers here, but if we did add this extra layer you could enforce those type of attributes you described.)
2. Instead of doing the previous, I compromised by adding `getSizeEstimate()` to `AbstractSymEntry` that all descendants would implement so we could easily estimate how much space was being used by things in the SymbolTable.  This worked out nicely in the SegStringSymEntry because it asks its internal components what their size estimates are and just adds them together.  It seemed like a nice way to encapsulate the internal bookeeping logic and made the mem-check estimates easier in the rest of the code.  I think this idea would work with storing arbitrary objects in the Symbol Table too.
3. I think the attributes you describe are mostly needed on the client end and I'm curious if we should add a `getDescription()` proc which could produce something more like a json formatted string which essentially lists out the attributes of the entry ... or make the `__str__` / `writeThis` methods do this.  (It's kind of like a mini-serialization of an object but only storing the meta-data attributes).  Currently the only place on the Server-side which needed `(dtype, entry.size, entry.itemsize)` was `ConcatenateMsg` as part of its logic so I created a `getArraySpecFromEntry` proc as a temporary work-around.

Hopefully this makes sense; it was combo of my thinking about what _might_ be stored in the SymbolTable in the future and what I discovered when trying to implement the changes.

Oh and I need to change the SymbolTableType names to match the actual Entry class names.  This would make `dclass` and `entryType` the same.

**Review by pierce314159 — APPROVED** [review]
Looks good to me!

File: arkouda/strings.py
Comment (glitch) [reply arkouda/strings.py:1152]: That was pre-existing, I think the whitespace around it changed.

File: arkouda/pdarrayIO.py
Comment (glitch) [reply arkouda/pdarrayIO.py:225]: No I don't think there is any reason other than PR drift over time.  This section of code has changed upstream a few times and I'm assuming one of my rebases wrapped the old constructor with the factory method instead of changing to the simpler version of _from_return_msg_

**Comment by reuster986** [comment]
> @reuster986 I know what you mean. It really hinges on whether or not we're going to allow arbitrary objects to be put in the SymbolTable i.e. non-array objects (obviously those wouldn't have `shape`, `size`, `dtype` etc.
> 
> I have a few thoughts on this:
> 
> 1. I toyed with the idea of having another layer in the hierarchy that designated things as "Arrays" (it would be a common parent of GenSymEntry & CompositeEntry) but it gets problematic with composites.  SegString has two arrays and they both have different lengths; we kind of squint at it now and use the offsets array as the "real" length, but as the list of Composites grow things like shape, length, dtype get weird.  Would Categorical & GroupBy have a dtype?  (I also got worried about adding too many layers here, but if we did add this extra layer you could enforce those type of attributes you described.)
> 2. Instead of doing the previous, I compromised by adding `getSizeEstimate()` to `AbstractSymEntry` that all descendants would implement so we could easily estimate how much space was being used by things in the SymbolTable.  This worked out nicely in the SegStringSymEntry because it asks its internal components what their size estimates are and just adds them together.  It seemed like a nice way to encapsulate the internal bookeeping logic and made the mem-check estimates easier in the rest of the code.  I think this idea would work with storing arbitrary objects in the Symbol Table too.
> 3. I think the attributes you describe are mostly needed on the client end and I'm curious if we should add a `getDescription()` proc which could produce something more like a json formatted string which essentially lists out the attributes of the entry ... or make the `__str__` / `writeThis` methods do this.  (It's kind of like a mini-serialization of an object but only storing the meta-data attributes).  Currently the only place on the Server-side which needed `(dtype, entry.size, entry.itemsize)` was `ConcatenateMsg` as part of its logic so I created a `getArraySpecFromEntry` proc as a temporary work-around.
> 
> Hopefully this makes sense; it was combo of my thinking about what _might_ be stored in the SymbolTable in the future and what I discovered when trying to implement the changes.
> 
> Oh and I need to change the SymbolTableType names to match the actual Entry class names. This would make `dclass` and `entryType` the same.

@glitch I think this all makes sense, and I agree with your reasoning. Well thought out!

**Review by reuster986 — APPROVED** [review]

**Comment by reuster986** [comment]
@mhmerrill I like the approach @glitch has taken here (introduce a class hierarchy that allows us to maintain one symbol table for all types of objects without changing GenSymEntry much), and I think we're ready to merge, but I wanted to see if you had any feedback first.

**Comment by mhmerrill** [comment]
@Bears-R-Us/arkouda-core-dev I am thinking we might merge this PR tomorrow. BUT @glitch has pushed changes since the last review, could someone who has reviewed this before @reuster986 or @pierce314159 re-review to make sure I am not doing something bad ;-)

**Comment by mhmerrill** [comment]
@Bears-R-Us/arkouda-core-dev since we had a couple of fixes merged today, I am going to wait until tomorrow to merge this.

**Pull request merged**
